src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs

```diff
@@ -72,19 +74,29 @@ private static List<IApiResponseMetadataProvider> GetResponseMetadataAttributes(
 
     private ICollection<ApiResponseType> GetApiResponseTypes(
        IReadOnlyList<IApiResponseMetadataProvider> responseMetadataAttributes,
+       IReadOnlyList<IProducesResponseTypeMetadata> producesResponseMetadata,
        Type? type,
        Type defaultErrorType)
     {
         var contentTypes = new MediaTypeCollection();
         var responseTypeMetadataProviders = _mvcOptions.OutputFormatters.OfType<IApiResponseTypeMetadataProvider>();
 
-        var responseTypes = ReadResponseMetadata(
+        // Read response metadata from providers
+        var responseTypesFromProvider = ReadResponseMetadata(
             responseMetadataAttributes,
             type,
             defaultErrorType,
             contentTypes,
             responseTypeMetadataProviders);
 
+        var responseTypesFromEndpointMetadata = ReadResponseMetadata(
+            producesResponseMetadata,
+            type,
+            responseTypeMetadataProviders,
+            _modelMetadataProvider);
+
+        var responseTypes = responseTypesFromProvider.Concat(responseTypesFromEndpointMetadata.Values).ToList();
```

Comment:
Which metadata do we prefer? In the case of duplicates it looks like the last one wins so we should concat the preferred list in order to iterate over it last.

Reply:
Or, should we be removing duplicates? We do return this list and use it elsewhere, so I don't know the implications of duplicate items

Reply:
It's additive. I think this is fine considering these are generally `IResult` implementations, so I hope MVC isn't inferring any other response types by default.

https://github.com/dotnet/aspnetcore/blob/8a4b4deb09c04134f22f8d39aae21d212282004f/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L341

---

> Or, should we be removing duplicates?

To go off on a bit of a tangent, this got me thinking about how we have to be careful dealing with multiple pieces of Endpoint metadata of the same type going forward. We can no longer assume that conventions will be able to "see" duplicates because of [the way route groups add metadata](https://github.com/dotnet/aspnetcore/blob/8a4b4deb09c04134f22f8d39aae21d212282004f/src/Http/Routing/src/GroupRouteBuilder.cs#L92-L115). Everything that wants to read additive metadata should be prepared for multiple pieces of metadata. This makes adding metadata easier, and I think more people are going to add metadata than read metadata.

See https://github.com/dotnet/aspnetcore/pull/41779 for an example of this. It's not really related to this PR but your comment got me thinking about this @BrennanConroy.

Reply:
I think it makes sense to prefer attributes applied to the action directly over metadata provided by the response type given a conflict for a given status code. Hopefully this never happens in practice.